lib/filterx/filterx-expr.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2023 Balazs Scheidler <balazs.scheidler@axoflow.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#ifndef FILTERX_EXPR_H_INCLUDED
+#define FILTERX_EXPR_H_INCLUDED
+
+#include "filterx-object.h"
+#include "template/eval.h"
+
+typedef struct _FilterXExpr FilterXExpr;
+typedef struct _FilterXEvalContext FilterXEvalContext;
+
+struct _FilterXEvalContext
+{
+  LogMessage **msgs;
+  gint num_msg;
+  LogTemplateEvalOptions *template_eval_options;
+};
+
+struct _FilterXExpr
+{
+  guint32 ref_cnt;
+  const gchar *type;
+  FilterXObject *(*eval)(FilterXExpr *self);
+  FilterXObject *(*eval_typed)(FilterXExpr *self);
```

Comment:
Why do we need a typed version of this function?

Reply:
The difference between those two is that the first one can return an unmarshalled representation of the object, the "typed" alternative would deserialize any external representation and return that.

This is a kind of "lazy" evaluation, because in some cases all you need is raw data.

Reply:
We've discussed and simplified this IRL and eval_typed() is now gone in favour of an _update_repr() method that is only implemented by expr-message-ref.c